File: frontend/src/contracts/inference.ts
import { z } from 'zod'
import {
  BucketSchema,
  CorpusDeviceIdSchema,
  HistoricalDateTimeSchema,
  SeveritySchema,
  compareHistoricalDateTimes,
} from './common'

export const ScoreProvenanceSchema = z.enum(['simulated_preview', 'artifact_backed'])

export const InferencePointSchema = z
  .strictObject({
    window_start_ts: HistoricalDateTimeSchema,
    window_end_ts: HistoricalDateTimeSchema,
    score_ts: HistoricalDateTimeSchema,
    score: z.number(),
    threshold: z.number(),
    is_anomaly: z.boolean(),
    severity: SeveritySchema.nullable(),
    latest_score: z.number().nullable(),
    sample_count: z.number().int().positive(),
    model_version: z.string(),
    score_provenance: ScoreProvenanceSchema,
    recon_temperature_c: z.number().nullable(),
    recon_relative_humidity_pct: z.number().nullable(),
    band_half_temperature_c: z.number().nonnegative().nullable(),
    band_half_relative_humidity_pct: z.number().nonnegative().nullable(),
  })
  .refine(
    (value) => compareHistoricalDateTimes(value.window_start_ts, value.window_end_ts) < 0,
    {
      message: 'window_start_ts must be earlier than window_end_ts',
      path: ['window_start_ts'],
    },
  )
  .refine(
    (value) => compareHistoricalDateTimes(value.window_end_ts, value.score_ts) <= 0,
    {
      message: 'score_ts must not be earlier than window_end_ts',
      path: ['score_ts'],
    },
  )
export type InferencePoint = z.infer<typeof InferencePointSchema>

export const InferenceQuerySchema = z
  .strictObject({
    deviceId: CorpusDeviceIdSchema,
    from: HistoricalDateTimeSchema,
    to: HistoricalDateTimeSchema,
    bucket: BucketSchema.default('raw'),
    limit: z.number().int().min(1).max(5_000).default(500),
    cursor: z.string().optional(),
    modelVersion: z.string().optional(),
  })
  .superRefine((value, context) => {
    if (compareHistoricalDateTimes(value.from, value.to) >= 0) {
      context.addIssue({ code: 'custom', message: 'from must be earlier than to', path: ['from'] })
    }
    if (value.bucket !== 'raw' && value.limit > 2_000) {
      context.addIssue({
        code: 'custom',
        message: 'bucketed limit must be at most 2000',
        path: ['limit'],
      })
    }
  })
export const InferenceResultsQuerySchema = InferenceQuerySchema
export type InferenceQuery = z.input<typeof InferenceQuerySchema>
export type InferenceResultsQuery = InferenceQuery

export const InferenceResponseSchema = z
  .strictObject({
    request_id: z.string(),
    device_id: CorpusDeviceIdSchema,
    from: HistoricalDateTimeSchema,
    to: HistoricalDateTimeSchema,
    bucket: BucketSchema,
    bucket_seconds: z.number().int().min(60).nullable(),
    time_zone: z.literal('Asia/Jakarta'),
    model_version: z.string(),
    points: z.array(InferencePointSchema).max(5_000),
    next_cursor: z.string().nullable(),
    returned_count: z.number().int().nonnegative(),
  })
  .superRefine((value, context) => {
    if (compareHistoricalDateTimes(value.from, value.to) >= 0) {
      context.addIssue({ code: 'custom', message: 'from must be earlier than to', path: ['from'] })
    }
    if (value.returned_count !== value.points.length) {
      context.addIssue({
        code: 'custom',
        message: 'returned_count must equal points length',
        path: ['returned_count'],
      })
    }
  })
  .refine(
    (value) => value.points.every((point) => point.model_version === value.model_version),
    {
      message: 'all points must belong to the selected model_version',
      path: ['points'],
    },
  )
export const InferenceResultsResponseSchema = InferenceResponseSchema
export type InferenceResponse = z.infer<typeof InferenceResponseSchema>
export type InferenceResultsResponse = InferenceResponse

File: backend/anomaly_backend/routes/inference.py
from datetime import datetime
from typing import Annotated, cast

from fastapi import APIRouter, Depends, Query
from sqlalchemy.engine import RowMapping
from sqlalchemy.ext.asyncio import AsyncConnection

from anomaly_backend.contracts import (
    Bucket,
    CorpusDeviceId,
    HistoricalDateTime,
    InferencePoint,
    InferenceQuery,
    InferenceResponse,
    ScoreProvenance,
    Severity,
    effective_bucket_seconds,
    format_historical_datetime,
    make_keyset_cursor,
    parse_keyset_cursor,
)
from anomaly_backend.db import get_connection
from anomaly_backend.problems import InvalidQuery, new_request_id
from anomaly_backend.sql import inference_rows


router = APIRouter()


def _datetime(row: RowMapping, field: str) -> datetime:
    return cast(datetime, row[field])


@router.get("/api/inference-results", response_model=InferenceResponse)
async def inference_results(
    device_id: Annotated[CorpusDeviceId, Query()],
    from_ts: Annotated[HistoricalDateTime, Query(alias="from")],
    to_ts: Annotated[HistoricalDateTime, Query(alias="to")],
    connection: Annotated[AsyncConnection, Depends(get_connection)],
    bucket: Annotated[Bucket, Query()] = "raw",
    limit: Annotated[int, Query(ge=1, le=5_000)] = 500,
    cursor: Annotated[str | None, Query()] = None,
    model_version: Annotated[str | None, Query()] = None,
) -> InferenceResponse:
    try:
        bucket_seconds = effective_bucket_seconds(bucket, from_ts, to_ts)
    except ValueError as error:
        raise InvalidQuery(
            "Query parameters failed validation",
            {"from": [str(error)]},
        ) from error
    if bucket != "raw" and limit > 2_000:
        raise InvalidQuery(
            "Query parameters failed validation",
            {"limit": ["bucketed limit must be at most 2000"]},
        )
    query_fields: dict[str, object] = {
        "device_id": device_id,
        "from": from_ts,
        "to": to_ts,
        "bucket": bucket,
        "limit": limit,
    }
    if cursor is not None:
        query_fields["cursor"] = cursor
    if model_version is not None:
        query_fields["model_version"] = model_version
    query = InferenceQuery.model_validate(query_fields, strict=True)
    filters: dict[str, object] = {
        "device_id": query.device_id,
        "from": query.from_ts,
        "bucket": query.bucket,
        "bucket_seconds": bucket_seconds,
        "model_version": query.model_version,
    }
    try:
        after = (
            parse_keyset_cursor(
                query.cursor,
                "inference",
                snapshot_to=query.to_ts,
                filters=filters,
            )
            if query.cursor
            else None
        )
    except ValueError as error:
        raise InvalidQuery(
            "Query parameters failed validation",
            {"cursor": ["Invalid cursor"]},
        ) from error
    model_version, rows = await inference_rows(
        connection,
        device_id=query.device_id,
        from_ts=datetime.fromisoformat(query.from_ts),
        to_ts=datetime.fromisoformat(query.to_ts),
        model_version=query.model_version,
        bucket=query.bucket,
        bucket_seconds=bucket_seconds,
        limit=query.limit,
        after_ts=datetime.fromisoformat(after[0]) if after else None,
        after_id=after[1] if after else None,
    )
    has_more = len(rows) > query.limit
    points = [
        InferencePoint(
            window_start_ts=format_historical_datetime(
                _datetime(row, "window_start_ts")
            ),
            window_end_ts=format_historical_datetime(_datetime(row, "window_end_ts")),
            score_ts=format_historical_datetime(_datetime(row, "score_ts")),
            score=cast(float, row["score"]),
            threshold=cast(float, row["threshold"]),
            is_anomaly=cast(bool, row["is_anomaly"]),
            model_version=cast(str, row["model_version"]),
            score_provenance=cast(ScoreProvenance, row["score_provenance"]),
            severity=cast(Severity, row["severity"]),
            latest_score=cast(float, row["latest_score"]),
            sample_count=cast(int, row["sample_count"]),
            recon_temperature_c=cast(float | None, row["recon_temperature_c"]),
            recon_relative_humidity_pct=cast(
                float | None, row["recon_relative_humidity_pct"]
            ),
            band_half_temperature_c=cast(
                float | None, row["band_half_temperature_c"]
            ),
            band_half_relative_humidity_pct=cast(
                float | None, row["band_half_relative_humidity_pct"]
            ),
        )
        for row in rows[: query.limit]
    ]
    return InferenceResponse.model_validate(
        {
            "request_id": new_request_id(),
            "device_id": query.device_id,
            "from": query.from_ts,
            "to": query.to_ts,
            "bucket": query.bucket,
            "bucket_seconds": bucket_seconds,
            "time_zone": "Asia/Jakarta",
            "model_version": model_version,
            "points": points,
            "next_cursor": (
                make_keyset_cursor(
                    "inference",
                    timestamp=format_historical_datetime(
                        _datetime(rows[query.limit - 1], "cursor_ts")
                    ),
                    row_id=cast(str, rows[query.limit - 1]["row_id"]),
                    snapshot_to=query.to_ts,
                    filters=filters,
                )
                if has_more
                else None
            ),
            "returned_count": len(points),
        },
        strict=True,
    )
